pkg/doc/verifiable/extension_credential_switch_test.go

```diff
@@ -0,0 +1,232 @@
+  ],
+  "id": "http://example.edu/credentials/1872",
+  "type": [
+    "VerifiableCredential",
+    "CredType2"
+  ],
+  "credentialSubject": {
+    "id": "did:example:ebfeb1f712ebc6f1c276e12ec21",
+    "s2": "custom subject 2"
+  },
+
+  "c2": "custom field 2",
+
+  "issuer": {
+    "id": "did:example:76e12ec712ebc6f1c221ebfeb1f",
+    "name": "Example University"
+  },
+
+  "issuanceDate": "2010-01-01T19:23:24Z"
+}`
+
+	credMissingMandatoryFields = `
+{
+  "@context": [
+    "https://www.w3.org/2018/credentials/v1",
+    "https://www.w3.org/2018/credentials/examples/ext/type2"
+  ],
+  "id": "http://example.edu/credentials/1872",
+  "type": [
+    "VerifiableCredential"
+  ]
+}`
+)
+
+type CustomCredentialProducer struct {
+	Accept func(cred *Credential) bool
+	Apply  func(dataJSON []byte) (interface{}, error)
+}
+
+type Cred1 struct {
+	Base    Credential
+	Subject struct {
+		ID                 string `json:"id,omitempty"`
+		CustomSubjectField string `json:"s1,omitempty"`
+	} `json:"credentialSubject,omitempty"`
+	CustomField string `json:"c1,omitempty"`
+}
+
+type Cred2 struct {
+	Base    Credential
+	Subject struct {
+		ID                 string `json:"id,omitempty"`
+		CustomSubjectField string `json:"s2,omitempty"`
+	} `json:"credentialSubject,omitempty"`
+	CustomField string `json:"c2,omitempty"`
+}
+
+func Cred1Producer() CustomCredentialProducer {
+	return CustomCredentialProducer{
+		Apply: func(dataJSON []byte) (interface{}, error) {
+			cred1 := &Cred1{}
+
+			_, err := NewCredential(
+				dataJSON,
+				WithDecoders([]CredentialDecoder{func(dataJSON []byte, c *Credential) error {
+					return json.Unmarshal(dataJSON, cred1)
+				}}),
+				WithTemplate(func() *Credential {
+					return &cred1.Base
+				}),
+			)
+			if err == nil {
+				return cred1, nil
+			}
+			return nil, err
+		},
+		Accept: func(vc *Credential) bool {
+			return containsString(vc.Context, "https://www.w3.org/2018/credentials/examples/ext/type1") &&
+				containsString(vc.Type, "CredType1")
+		},
+	}
+}
+
+func Cred2Producer() CustomCredentialProducer {
+	return CustomCredentialProducer{
+		Apply: func(dataJSON []byte) (interface{}, error) {
+			cred2 := &Cred2{}
+
+			_, err := NewCredential(
+				dataJSON,
+				WithDecoders([]CredentialDecoder{func(dataJSON []byte, c *Credential) error {
+					return json.Unmarshal(dataJSON, cred2)
+				}}),
+				WithTemplate(func() *Credential {
+					return &cred2.Base
+				}),
+			)
+			if err == nil {
+				return cred2, nil
+			}
+			return nil, err
+		},
+		Accept: func(vc *Credential) bool {
+			return containsString(vc.Context, "https://www.w3.org/2018/credentials/examples/ext/type2") &&
+				containsString(vc.Type, "CredType2")
+		},
+	}
+}
+
+func DecodeCredentials(producers ...CustomCredentialProducer) (chan<- []byte, <-chan interface{}, <-chan error) {
+	input := make(chan []byte)
+	output := make(chan interface{})
+	errors := make(chan error)
+
+	go func() {
```

Comment:
Why a new goroutine?

Reply:
Might be just for fun... Seriously, we can use function call here.

Reply:
As an example, we should show the simplest version that helps the reader understand - in this case understand how to use a basic credential to create a customized one.

Reply:
Simplified! And now it actually looks much clear and compact - so thanks for the suggestion @troyronda !